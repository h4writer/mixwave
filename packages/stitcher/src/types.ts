export type Session = {
  id: string;
  assetId: string;
  interstitials: Interstitial[];
};

export type Interstitial = {
  timeOffset: number;
  assetId: string;
};
