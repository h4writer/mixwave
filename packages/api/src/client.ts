export { streamSchema, inputSchema } from "@mixwave/artisan/schemas";

export * from "./contract";

export type { JobDto } from "./types";
