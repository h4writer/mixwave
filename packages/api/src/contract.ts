import { initContract } from "@ts-rest/core";
import { streamSchema, inputSchema } from "@mixwave/artisan/schemas";
import * as z from "zod";
import type { JobDto } from "./types.js";

const c = initContract();

export const postTranscodeBodySchema = z.object({
  inputs: z.array(inputSchema),
  streams: z.array(streamSchema),
  segmentSize: z.number().default(4),
  assetId: z.string().uuid().optional(),
  package: z.boolean().default(false),
  tag: z.string().default("default"),
});

export const postPackageBodySchema = z.object({
  assetId: z.string(),
  segmentSize: z.number().default(4),
  tag: z.string().default("default"),
});

export const contract = c.router({
  postTranscode: {
    method: "POST",
    path: "/transcode",
    body: postTranscodeBodySchema,
    responses: {
      200: c.type<{ jobId: string }>(),
    },
    description: "Convert your source files and prepare them for packaging.",
  },
  postPackage: {
    method: "POST",
    path: "/package",
    body: postPackageBodySchema,
    responses: {
      200: c.type<{ jobId: string }>(),
    },
    description: "Package your asset to HLS.",
  },
  getJobs: {
    method: "GET",
    path: "/jobs",
    responses: {
      200: c.type<JobDto[]>(),
    },
  },
  getJob: {
    method: "GET",
    path: "/jobs/:id",
    responses: {
      200: c.type<JobDto>(),
    },
    query: z.object({
      fromRoot: z.coerce.boolean().default(false),
    }),
  },
  getJobLogs: {
    method: "GET",
    path: "/jobs/:id/logs",
    responses: {
      200: c.type<string[]>(),
    },
  },
  getSpec: {
    method: "GET",
    path: "/spec.json",
    responses: {},
  },
});
