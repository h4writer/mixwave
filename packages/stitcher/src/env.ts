import { z } from "zod";

const envSchema = z.object({
  PORT: z.coerce.number(),
  HOST: z.string(),
  REDIS_HOST: z.string(),
  REDIS_PORT: z.coerce.number(),
  S3_PUBLIC_URL: z.string(),
});

export const env = envSchema.parse(process.env);
