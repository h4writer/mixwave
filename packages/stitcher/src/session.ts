import { client } from "./redis.js";
import { randomUUID } from "crypto";
import { getInterstitialsFromVmap } from "./vmap.js";
import type { Session, Interstitial } from "./types.js";

const REDIS_PREFIX = `stitcher:session`;

const key = (sessionId: string) => `${REDIS_PREFIX}:${sessionId}`;

export async function createSession(data: {
  assetId: string;
  vmapUrl?: string;
  interstitials?: Interstitial[];
  bumperAssetId?: string;
}) {
  const sessionId = randomUUID();

  const interstitials: Interstitial[] = [];

  if (data.vmapUrl) {
    interstitials.push(...(await getInterstitialsFromVmap(data.vmapUrl)));
  }
  if (data.interstitials) {
    interstitials.push(...data.interstitials);
  }

  // When we have a bumper, we'll push it at the end of the interstitials list.
  if (data.bumperAssetId) {
    interstitials.push({
      timeOffset: 0,
      assetId: data.bumperAssetId,
    });
  }

  const session = {
    id: sessionId,
    assetId: data.assetId,
    interstitials,
  } satisfies Session;

  await client.json.set(key(sessionId), `$`, session);

  await client.expire(key(sessionId), 60 * 60 * 6);

  return session;
}

export async function getSession(sessionId: string) {
  const data = await client.json.get(key(sessionId));
  if (!data) {
    throw new Error("No session found for id");
  }
  return data as Session;
}
