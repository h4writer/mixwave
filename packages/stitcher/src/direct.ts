import base64 from "hi-base64";
import { postSessionBodySchema } from "./contract.js";

export function parseDirectParams(value: string) {
  const raw = base64.decode(value);
  const obj = JSON.parse(raw);
  return postSessionBodySchema.parse(obj);
}
