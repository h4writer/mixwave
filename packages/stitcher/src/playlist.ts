import { parse, stringify } from "./extern/hls-parser/index.js";
import parseFilepath from "parse-filepath";
import { Interstitial } from "./extern/hls-parser/types.js";
import { env } from "./env.js";
import { MasterPlaylist, MediaPlaylist } from "./extern/hls-parser/types.js";
import type { Session } from "./types.js";

async function fetchPlaylist<T>(url: string) {
  const response = await fetch(url);
  const text = await response.text();
  return parse(text) as T;
}

export async function formatMasterPlaylist(session: Session) {
  const url = `${env.S3_PUBLIC_URL}/package/${session.assetId}/hls/master.m3u8`;

  const master = await fetchPlaylist<MasterPlaylist>(url);

  return stringify(master);
}

export async function formatMediaPlaylist(session: Session, path: string) {
  const url = `${env.S3_PUBLIC_URL}/package/${session.assetId}/hls/${path}/playlist.m3u8`;

  const media = await fetchPlaylist<MediaPlaylist>(url);

  const filePath = parseFilepath(url);

  for (const segment of media.segments) {
    if (segment.map?.uri === "init.mp4") {
      segment.map.uri = `${filePath.dir}/init.mp4`;
    }

    segment.uri = `${filePath.dir}/${segment.uri}`;
  }

  const now = Date.now();

  media.segments[0].programDateTime = new Date(now);

  session.interstitials
    .reduce<number[]>((acc, interstitial) => {
      if (!acc.includes(interstitial.timeOffset)) {
        acc.push(interstitial.timeOffset);
      }
      return acc;
    }, [])
    .forEach((timeOffset) => {
      media.interstitials.push(
        new Interstitial({
          id: `${timeOffset}`,
          startDate: new Date(now + timeOffset * 1000),
          list: `/session/${session.id}/asset-list.json?timeOffset=${timeOffset}`,
        }),
      );
    });

  return stringify(media);
}

export async function formatAssetList(session: Session, timeOffset: number) {
  const interstitials = session.interstitials.filter(
    (ad) => ad.timeOffset === timeOffset,
  );

  const assets = await Promise.all(
    interstitials.map(async (interstitial) => {
      const uri = `${env.S3_PUBLIC_URL}/package/${interstitial.assetId}/hls/master.m3u8`;
      return {
        URI: uri,
        DURATION: await getDuration(uri),
      };
    }),
  );

  return { ASSETS: assets };
}

async function getDuration(url: string) {
  const master = await fetchPlaylist<MasterPlaylist>(url);
  const filePath = parseFilepath(url);
  const media = await fetchPlaylist<MediaPlaylist>(
    `${filePath.dir}/${master.variants[0].uri}`,
  );
  return media.segments.reduce((acc, segment) => {
    acc += segment.duration;
    return acc;
  }, 0);
}
