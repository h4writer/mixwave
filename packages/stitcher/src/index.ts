import Fastify from "fastify";
import cors from "@fastify/cors";
import { env } from "./env.js";
import { contract } from "./contract.js";
import { initServer } from "@ts-rest/fastify";
import { generateOpenApi } from "@ts-rest/open-api";
import { createSession, getSession } from "./session.js";
import {
  formatMasterPlaylist,
  formatMediaPlaylist,
  formatInterstitialsJson,
} from "./playlist.js";
import { parseDirectParams } from "./direct.js";

async function buildServer() {
  const app = Fastify();

  app.register(cors);

  const s = initServer();

  const router = s.router(contract, {
    postSession: async ({ request, body }) => {
      const session = await createSession(body);

      return {
        status: 200,
        body: {
          url: `${request.protocol}://${request.hostname}/session/${session.id}/master.m3u8`,
          session,
        },
      };
    },
    getMasterPlaylist: async ({ params, reply }) => {
      const session = await getSession(params.sessionId);
      const response = await formatMasterPlaylist(session);

      return reply.type("application/x-mpegURL").send(response);
    },
    getMediaPlaylist: async ({ params, reply }) => {
      const session = await getSession(params.sessionId);
      const response = await formatMediaPlaylist(session, params.path);

      return reply.type("application/x-mpegURL").send(response);
    },
    getInterstitialsList: async ({ query, params }) => {
      const session = await getSession(params.sessionId);
      return {
        status: 200,
        body: await formatInterstitialsJson(session, query.offset),
      };
    },
    getSpec: async () => {
      return {
        status: 200,
        body: generateOpenApi(contract, {
          info: {
            title: "Stitcher",
            version: "1.0.0",
          },
        }),
      };
    },
    unstable_getDirectMasterPlaylist: async ({ request, query, reply }) => {
      const body = parseDirectParams(query.params);
      const session = await createSession(body);

      return reply.redirect(
        `${request.protocol}://${request.hostname}/session/${session.id}/master.m3u8`,
        302,
      );
    },
  });

  app.register(s.plugin(router));

  return app;
}

async function main() {
  const app = await buildServer();

  await app.listen({ port: env.PORT });
}

main();
