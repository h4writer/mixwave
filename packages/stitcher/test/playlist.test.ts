import "./env-mock";
import fetchMock from "fetch-mock";
import { formatMasterPlaylist } from "../src/playlist";
import { describe, test, expect } from "@jest/globals";

describe("playlist", () => {
  test("formatMasterPlaylist", async () => {
    fetchMock.mock("https://s3-public.com/package/mock/hls/master.m3u8", {
      status: 200,
      body: `
        #EXTM3U
        #EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=2149280,CODECS="mp4a.40.2,avc1.64001f",RESOLUTION=1280x720,NAME="720" 
        720.m3u8
      `,
    });

    const master = await formatMasterPlaylist({
      id: "mock",
      assetId: "mock",
      interstitials: [],
    });

    expect(master).toMatchSnapshot();
  });
});
