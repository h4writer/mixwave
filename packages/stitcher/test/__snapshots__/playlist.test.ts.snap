// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`playlist formatMasterPlaylist 1`] = `
"#EXTM3U
#EXT-X-STREAM-INF:BANDWIDTH=2149280,CODECS="mp4a.40.2,avc1.64001f",RESOLUTION=1280x720,PROGRAM-ID=1
720.m3u8"
`;
